src/app/directives/has-permission/has-permission.directive.ts

```diff
@@ -0,0 +1,78 @@
+/** Angular Imports */
+import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
+
+/**
+ * Has Permission Directive
+ */
+@Directive({
+  selector: '[mifosxHasPermission]'
+})
+export class HasPermissionDirective {
+
+  /** Describes Visibility */
+  private isVisible: boolean;
+  /** Flag if permission passed is negative */
+  private isNegation = false;
+  /** User Permissions */
+  private userPermissions: any[];
+  /** Key to store credentials in storage. */
+  private credentialsStorageKey = 'mifosXCredentials';
+
+  /**
+   * Extracts User Permissions from User Credentials
+   * @param {TemplateRef} templateRef Template Reference
+   * @param {ViewContainerRef} viewContainer View Container Reference
+   */
+  constructor(private templateRef: TemplateRef<any>,
+              private viewContainer: ViewContainerRef) {
+    const savedCredentials = JSON.parse(
+      sessionStorage.getItem(this.credentialsStorageKey) || localStorage.getItem(this.credentialsStorageKey)
+    );
+    this.userPermissions = savedCredentials.permissions;
+  }
+
+  /**
+   * Evaluates the condition to show template.
+   */
+  @Input()
+  set mifosxHasPermission(permission: any) {
+    if (typeof permission !== 'string') {
+      throw new Error('hasPermission value must be a string');
+    }
+    if (permission.charAt(0) === '!') {
+      this.isNegation = true;
+      permission = permission.slice(1);
+    }
+    if (this.hasPermission(permission) && !this.isNegation) {
+      this.viewContainer.createEmbeddedView(this.templateRef);
+      this.isVisible = true;
+    } else {
+      this.viewContainer.clear();
+      this.isVisible = false;
+    }
+  }
+
+  /**
+   * Checks if user is permitted.
+   * @param {string} permission Permission
+   * @returns {true} for 'all functions', 'all function read', and special user permissions.
+   * @returns {false} for null/undefined permission or lack of permission.
+   */
+  private hasPermission(permission: string) {
+    permission = permission.trim();
+    if (this.userPermissions.includes('ALL_FUNCTIONS')) {
+      return true;
+    } else if (permission !== null || undefined) {
```

Comment:
```suggestion
    } else if (permission !== null || undefined) {
```
undefined is not compared with any parameter.

Reply:
oops